### PR TITLE
Values escaped in browse and discover links

### PR DESCRIPTION
**Pull request by MarieVerdonck**
## References
* Part of fix for https://github.com/DSpace/DSpace/issues/3230 ('Browse by author broken if author name has special character'), but will require an additional Angular fix
* Angular fix: https://github.com/DSpace/dspace-angular/pull/1216 ('Escape browse by author data requests')

## Description
Escapes the links in the browse entries items link and in the discovery search link so that spaces, ampersands, etc are escaped in these links so they correctly resolve with results:
* `{url}/api/discover/search/objects` => `_embedded.facets._embedded.values._links.search.href`
* `{url}/api/discover/browses/author/entries` => `_embedded.entries._links.items.href`

_Example:_ 
Before:
```
"items": {
  "href": "{{url}}/server/api/discover/browses/author/items?filterValue=Biomolecular & Analytical Mass Spectrometry (BAMS)"
```
After: 
```
"items": {
  "href": "{{url}}/server/api/discover/browses/author/items?filterValue=Biomolecular%20%26%20Analytical%20Mass%20Spectrometry%20(BAMS)"
```

## Instructions for Reviewers
- The browse entries request done in the frontend must result in entries with an items link which is escaped.
So for _example_ with https://demo7.dspace.org/browse/author?bbm.rpp=100&startsWith=Biomolecular%20%26
=> Does request https://api7.dspace.org/server/api/discover/browses/author/entries?sort=default,ASC&page=0&size=100&startsWith=Biomolecular%20%26 whose resulting `_embedded.entries._links.items.href` should be escaped like described above.
=> Verify this items link has [results](https://api7.dspace.org/server/api/discover/browses/author/items?filterValue=Biomolecular%20%26%20Analytical%20Mass%20Spectrometry%20(BAMS))

- However the actual author browse page https://demo7.dspace.org/browse/author?bbm.rpp=100&value=Biomolecular%20%26%20Analytical%20Mass%20Spectrometry%20(BAMS) still doesn't show the (4) expected items => Requests https://api7.dspace.org/server/api/discover/browses/author/items?sort=default,ASC&page=0&size=100&filterValue=Biomolecular%20&%20Analytical%20Mass%20Spectrometry%20(BAMS) => Ampersand not escaped
Cause: The frontend builds the link again in `BrowseEntrySearchOptions#getBrowseItemsFor(filterValue, BrowseEntrySearchOptions)` where the filterValue should also be escaped [here](https://github.com/DSpace/dspace-angular/blob/main/src/app/core/browse/browse.service.ts#L133) 
=> Requires additional minor Angular change to fix the original issue

- The discovery search request from the discovery page (ex  https://demo7.dspace.org/search?value=&spc.sf=score&spc.sd=DESC&spc.page=1&query=Test&scope= => https://api7.dspace.org/server/api/discover/search/objects?sort=score,DESC&page=0&size=10&query=Test) in its response in the `_embedded.facets._embedded.values._links.search.href` the values should be escaped
_Before_:
```
                "search": {
                  "href": "https://api7.dspace.org/server/api/discover/search/objects?query=Test&f.author=Simmons, Cameron,equals"
```
_After_:
```
                "search": {
                  "href": "https://api7.dspace.org/server/api/discover/search/objects?query=Test&f.author=Simmons,%20Cameron,equals"
```

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

**Review by abollini — APPROVED** [review]
thanks @MarieVerdonck code looks good and IT proof that it works

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/DiscoveryRestControllerIT.java [hunk 1]
```diff
@@ -1140,6 +1142,58 @@ public void discoverSearchObjectsTest() throws Exception {
...
 
+    @Test
+    public void discoverSearchObjectsWithSpecialCharacterTest() throws Exception {
+
+        context.turnOffAuthorisationSystem();
+
+        parentCommunity = CommunityBuilder.createCommunity(context).build();
+        Collection collection = CollectionBuilder.createCollection(context, parentCommunity).build();
+        ItemBuilder.createItem(context, collection)
+                .withAuthor("Atmire & friends")
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/DiscoveryRestControllerIT.java:1153]: Tiny thing, but I'd recommend keeping test names entirely generic (as I don't want each service provider getting in the habit of inserting their name in our shared tests).  Could we change this to say "DSpace & friends" or give a generic person name like "Smith, Jane"?

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/DiscoveryRestControllerIT.java [hunk 2]
```diff
@@ -1140,6 +1142,58 @@ public void discoverSearchObjectsTest() throws Exception {
...
+
+    @Test
+    public void discoverSearchBrowsesWithSpecialCharacterTest() throws Exception {
+
+        context.turnOffAuthorisationSystem();
+
+        parentCommunity = CommunityBuilder.createCommunity(context).build();
+        Collection collection = CollectionBuilder.createCollection(context, parentCommunity).build();
+        ItemBuilder.createItem(context, collection)
+                .withAuthor("Atmire & friends")
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/DiscoveryRestControllerIT.java:1180]: Same as above,  I'd recommend keeping test names entirely generic.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@MarieVerdonck & @samuelcambien  : Got a chance to look at this quickly & it looks correct and has tests to prove it works.  However, I'd appreciate a *very minor* change to your tests (see inline comments below) before we merge this.  Thanks!

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/DiscoveryRestControllerIT.java (hunk 1 above)
Comment (samuelcambien) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/DiscoveryRestControllerIT.java]: Hi Tim, I understand, I've changed it to 'DSpace & friends'.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/DiscoveryRestControllerIT.java (hunk 2 above)
Comment (samuelcambien) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/DiscoveryRestControllerIT.java]: Hi Tim, I understand, I've changed it to 'DSpace & friends'.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @samuelcambien and @MarieVerdonck ! This all looks good now.

**Pull request merged**